src/xrpld/app/tx/detail/OnChainCredentials.cpp

```diff
@@ -0,0 +1,454 @@
+        // either case the transaction is malformed.
+        JLOG(j.trace())
+            << "Malformed transaction: "
+               "Invalid Subject, Issuer and Signature fields combination.";
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    AccountID const target{optSubj ? *optSubj : *optIss};
+    if (!target)
+    {
+        JLOG(j.trace()) << "Malformed transaction: Subject or Issuer "
+                           "field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    auto const optUri = tx[~sfURI];
+    if (optUri && (optUri->size() > maxCredentialURILength))
+    {
+        JLOG(j.trace()) << "Malformed transaction: "
+                           "URI too long.";
+        return temMALFORMED;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialCreate::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const acc{ctx.tx[sfAccount]};
+
+    auto const& txCredType(ctx.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    if (ctx.tx.isFieldPresent(sfSubject))
+    {
+        AccountID const subj{ctx.tx[sfSubject]};
+        if (!ctx.view.exists(keylet::account(subj)))
+            return tecNO_TARGET;
+        if (ctx.view.exists(keylet::credential(subj, acc, credType)))
+            return tecDUPLICATE;
+    }
+    else if (ctx.tx.isFieldPresent(sfIssuer))
+    {
+        AccountID const issuer{ctx.tx[sfIssuer]};
+        if (!ctx.view.exists(keylet::account(issuer)))
+            return tecNO_TARGET;
+        if (ctx.view.exists(keylet::credential(acc, issuer, credType)))
+            return tecDUPLICATE;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialCreate::doApply()
+{
+    std::uint32_t const closeTime =
+        ctx_.view().info().parentCloseTime.time_since_epoch().count();
+
+    auto const sleOwner = view().peek(keylet::account(account_));
+    if (!sleOwner)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        STAmount const reserve{view().fees().accountReserve(
+            sleOwner->getFieldU32(sfOwnerCount) + 1)};
+
+        if (mPriorBalance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    auto const& txCredType(ctx_.tx[sfCredentialType]);
+    Blob const credType{txCredType.begin(), txCredType.end()};
+
+    AccountID const subj = ctx_.tx.isFieldPresent(sfSubject)
+        ? AccountID(ctx_.tx[sfSubject])
+        : account_;
+    bool const issPresent = ctx_.tx.isFieldPresent(sfIssuer);
+    AccountID const issuer =
+        issPresent ? AccountID(ctx_.tx[sfIssuer]) : account_;
+
+    Keylet const kCred = keylet::credential(subj, issuer, credType);
+
+    auto const sleCred = std::make_shared<SLE>(kCred);
+    sleCred->setAccountID(sfSubject, subj);
+    sleCred->setAccountID(sfIssuer, issuer);
+    sleCred->setFieldVL(sfCredentialType, credType);
+
+    if (ctx_.tx.isFieldPresent(sfSignature))
```

Comment:
Shouldn't the signature be verified? And it shouldn't be included in the object according to the specs.

Reply:
Fixed. `sfIssuer` field in CreteCredential transaction changed to `sfPublicKey` field (which should be public key of the issuer for checking `Signature`)